src/api/forms/service.js

```diff
@@ -1,25 +1,140 @@
+import { readFile } from 'node:fs/promises'
+
+import { Schema } from '@defra/forms-model'
+
+import { createLogger } from './../../helpers/logging/logger.js'
+import {
+  FailedCreationOperationError,
+  FormAlreadyExistsError,
+  InvalidFormDefinitionError
+} from './errors.js'
+import {
+  createFormDefinition,
+  getFormDefinition as getFormDefinitionFromRepository
+} from './form-definition-repository.js'
 import {
+  createFormMetadata,
   listForms as listFormMetadataEntries,
+  exists as formMetadataExists,
   getFormMetadata
 } from './form-metadata-repository.js'
 
+const logger = createLogger()
+
+/**
+ * Adds an empty form
+ * @param {FormConfigurationInput} formConfigurationInput - the desired form configuration to save
+ * @returns {Promise<FormConfiguration>} - the saved form configuration
+ * @throws {FormAlreadyExistsError} - if the form already exists
+ * @throws {InvalidFormDefinitionError} - if the form definition is invalid
+ * @throws {FailedCreationOperationError} - if the form metadata/def couldn't be persisted
+ */
+export async function createForm(formConfigurationInput) {
+  const emptyForm = await retrieveEmptyForm()
+  const formId = formTitleToId(formConfigurationInput.title)
+
+  if (await formMetadataExists(formId)) {
+    throw new FormAlreadyExistsError(formId)
+  }
+
+  // construct the new form config. the ID is always set server-side.
+  const formConfiguration = { ...formConfigurationInput, id: formId }
+
+  // create the form object. At this point, we're just creating a blank
+  // form following the one-page template, we just set the title per the user request.
+  const shallowCloneForm = { ...emptyForm, name: formConfiguration.title }
+
+  const validationResult = Schema.validate(shallowCloneForm)
+  if (validationResult.error) {
+    throw new InvalidFormDefinitionError(validationResult.error.toString())
+  }
+
+  try {
+    await createFormDefinition(formConfiguration, shallowCloneForm)
+    await createFormMetadata(formConfiguration)
+  } catch (error) {
+    logger.error(error, "Failed to persist, couldn't create form.")
+    throw new FailedCreationOperationError()
+  }
```

Comment:
I intentionally left metadata as the last operation, since that's the 'glue' that makes it visible to the other query endpoints. A missing metadata entry means the form doesn't exist, regardless of whether the definition was written or not. This way, the user can at least re-try it in the event of an error and it will recover (if a transient error) - it'll just overwrite the orphaned form definition.

We'll be bringing in Mongo soon, that's probably a good point for us to make any improvements.

Reply:
>…it'll just overwrite the orphaned form definition

Cheers, that's answered it